asyncpg/protocol/protocol.pyx

```diff
@@ -332,6 +336,107 @@ cdef class BaseProtocol(CoreProtocol):
+            self.cancel_sent_waiter = None
+
+        self._check_state()
+
+        timeout = self._get_timeout_impl(timeout)
+        timer = Timer(timeout)
+
+        waiter = self._new_waiter(timer.get_remaining_budget())
+
+        # Initiate COPY IN.
+        self._copy_in(copy_stmt)
+
+        try:
+            if record_stmt is not None:
+                # copy_in_records in binary mode
+                wbuf = WriteBuffer.new()
+                # Signature
+                wbuf.write_bytes(_COPY_SIGNATURE)
+                # Flags field
+                wbuf.write_int32(0)
+                # No header extension
+                wbuf.write_int32(0)
+
+                record_stmt._ensure_rows_decoder()
+                codecs = record_stmt.rows_codecs
+                num_cols = len(codecs)
+                settings = self.settings
+
+                for codec in codecs:
+                    if not codec.has_encoder():
+                        raise RuntimeError(
+                            'no encoder for OID {}'.format(codec.oid))
+
+                while True:
+                    for row in records:
+                        # Tuple header
+                        wbuf.write_int16(<int16_t>num_cols)
+                        # Tuple data
+                        for i in range(num_cols):
+                            codec = <Codec>cpython.PyTuple_GET_ITEM(codecs, i)
+                            codec.encode(settings, wbuf, row[i])
+
+                        if wbuf.len() >= _COPY_BUFFER_SIZE:
+                            self._write_copy_data_msg(wbuf)
+                            wbuf = WriteBuffer.new()
+                            break
+                    else:
+                        # End of binary copy.
+                        wbuf.write_int16(-1)
+                        self._write_copy_data_msg(wbuf)
+                        break
+
+            elif reader is not None:
+                iterator = reader.__aiter__()
+
+                try:
+                    while True:
+                        # We rely on protocol flow control to moderate the
+                        # rate of data messages.
+                        with timer:
+                            await self.writing_allowed.wait()
+                        with timer:
+                            chunk = await asyncio.wait_for(
+                                iterator.__anext__(),
+                                timeout=timer.get_remaining_budget(),
+                                loop=self.loop)
+                        self._write_copy_data_msg(chunk)
+                except builtins.StopAsyncIteration:
+                    pass
+            else:
+                # Buffer passed in directly.
+                await self.writing_allowed.wait()
+                self._write_copy_data_msg(data)
+
+        except asyncio.TimeoutError:
+            self._write_copy_fail_msg('TimeoutError')
+            self._on_timeout(self.waiter)
+            await waiter
```

Comment:
Maybe instead of `await waiter` do this:

```python
try:
  await waiter
except TimeoutError:
  raise
else:
  raise RuntimeError('TimoutError was not raised')
```